Furball.Vixie.Backends.Direct3D11/QuadRendererD3D11.cs

```diff
@@ -1,297 +1,343 @@
+            this._instanceData[this._instances].InstancePosition              = position;
+            this._instanceData[this._instances].InstanceSize                  = texture.Size * scale;
+            this._instanceData[this._instances].InstanceColor                 = colorOverride;
+            this._instanceData[this._instances].InstanceRotation              = rotation;
+            this._instanceData[this._instances].InstanceRotationOrigin        = rotOrigin;
+            this._instanceData[this._instances].InstanceTextureId             = this.GetTextureId(texture);
+            this._instanceData[this._instances].InstanceTextureRectPosition.X = 0;
+            this._instanceData[this._instances].InstanceTextureRectPosition.Y = 0;
+            this._instanceData[this._instances].InstanceTextureRectSize.X     = texFlip == TextureFlip.FlipHorizontal ? -1 : 1;
+            this._instanceData[this._instances].InstanceTextureRectSize.Y     = texFlip == TextureFlip.FlipVertical ? -1 : 1;
+
+            this._instances++;
         }
         public void Draw(Texture textureGl, Vector2 position, Vector2 scale, float rotation, Color colorOverride, Rectangle sourceRect, TextureFlip texFlip = TextureFlip.None, Vector2 rotOrigin = default) {
-            fixed (VertexData* ptr = this._localVertexBuffer)
-                this._vertexBufferPointer = ptr;
-
-            Vector2 size = new Vector2(sourceRect.Width, sourceRect.Height) * scale;
-
-            Vector2 topLeftUv = Vector2.Zero;
-            Vector2 bottomRightUv = Vector2.Zero;
-
-            switch (texFlip) {
-                case TextureFlip.None:
-                    topLeftUv     = new Vector2(sourceRect.X                      * (1.0f / textureGl.Width), 1 - (sourceRect.Y + sourceRect.Height) * (1.0f / textureGl.Height));
-                    bottomRightUv = new Vector2((sourceRect.X + sourceRect.Width) * (1.0f / textureGl.Width), 1 - sourceRect.Y                       * (1.0f / textureGl.Height));
-                    break;
-                case TextureFlip.FlipVertical:
-                    topLeftUv     = new Vector2(sourceRect.X                      * (1.0f / textureGl.Width), 1 - sourceRect.Y                       * (1.0f / textureGl.Height));
-                    bottomRightUv = new Vector2((sourceRect.X + sourceRect.Width) * (1.0f / textureGl.Width), 1 - (sourceRect.Y + sourceRect.Height) * (1.0f / textureGl.Height));
-                    break;
-                case TextureFlip.FlipHorizontal:
-                    topLeftUv     = new Vector2((sourceRect.X + sourceRect.Width) * (1.0f / textureGl.Width), 1 - (sourceRect.Y + sourceRect.Height) * (1.0f / textureGl.Height));
-                    bottomRightUv = new Vector2(sourceRect.X                      * (1.0f / textureGl.Width), 1 - sourceRect.Y                       * (1.0f / textureGl.Height));
-                    break;
-            }
+            if (textureGl == null || textureGl is not TextureD3D11 texture)
```

Comment:
Should prob change `textureGl` to `texture`

Reply:
yeah